modules/wechat_qrcode/src/wechat_qrcode.cpp

```diff
@@ -0,0 +1,210 @@
+    std::shared_ptr<SSDDetector> detector_;
+    std::shared_ptr<SuperScale> super_resolution_model_;
+    bool use_nn_detector_, use_nn_sr_;
+};
+
+WeChatQRCode::WeChatQRCode(const String& detector_prototxt_path,
+                           const String& detector_caffe_model_path,
+                           const String& super_resolution_prototxt_path,
+                           const String& super_resolution_caffe_model_path) {
+    p = makePtr<WeChatQRCode::Impl>();
+    if (!detector_caffe_model_path.empty() && !detector_prototxt_path.empty()) {
+        // initialize detector model (caffe)
+        p->use_nn_detector_ = true;
+        CV_CheckEQ(utils::fs::exists(detector_prototxt_path), true,
+                   "fail to find detector caffe prototxt file");
+        CV_CheckEQ(utils::fs::exists(detector_caffe_model_path), true,
+                   "fail to find detector caffe model file");
+        p->detector_ = make_shared<SSDDetector>();
+        auto ret = p->detector_->init(detector_prototxt_path, detector_caffe_model_path);
+        CV_CheckEQ(ret, 0, "fail to load the detector model.");
+    } else {
+        p->use_nn_detector_ = false;
+        p->detector_ = NULL;
+    }
+    // initialize super_resolution_model
+    // it could also support non model weights by cubic resizing
+    // so, we initialize it first.
+    p->super_resolution_model_ = make_shared<SuperScale>();
+    if (!super_resolution_prototxt_path.empty() && !super_resolution_caffe_model_path.empty()) {
+        p->use_nn_sr_ = true;
+        // initialize dnn model (onnx format)
+        CV_CheckEQ(utils::fs::exists(super_resolution_prototxt_path), true,
+                   "fail to find super resolution prototxt model file");
+        CV_CheckEQ(utils::fs::exists(super_resolution_caffe_model_path), true,
+                   "fail to find super resolution caffe model file");
+        auto ret = p->super_resolution_model_->init(super_resolution_prototxt_path,
+                                                    super_resolution_caffe_model_path);
+        CV_CheckEQ(ret, 0, "fail to load the super resolution model.");
+    } else {
+        p->use_nn_sr_ = false;
+    }
+}
+
+vector<string> WeChatQRCode::detectAndDecode(InputArray img, OutputArrayOfArrays points) {
+    CV_Assert(!img.empty());
+    CV_CheckDepthEQ(img.depth(), CV_8U, "");
+
+    if (img.cols() <= 20 || img.rows() <= 20) {
+        return vector<string>();  // image data is not enough for providing reliable results
+    }
+    Mat input_img;
+    int incn = img.channels();
+    CV_Check(incn, incn == 1 || incn == 3 || incn == 4, "");
+    if (incn == 3 || incn == 4) {
+        cv::cvtColor(img, input_img, cv::COLOR_BGR2GRAY);
+    } else {
+        input_img = img.getMat();
+    }
+    auto candidate_points = p->detect(input_img);
+    auto res_points = vector<Mat>();
+    auto ret = p->decode(input_img, candidate_points, res_points);
```

Comment:
Are points fixed size? 4 points per instance?

if so, then it make sense to return 2D Mat instead of vector:
- rows are instances
- cols(=4) are corners points of `CV_32FC2` elements (C2 means 2 channels for x/y components)

Reply:
The size of points is not fixed. Because our detector supports detecting multiple QRcode in an image. Moreover, not all detected QRCode can be decoded. So, there are `candidate_points` detected by the detector. Then we try to decode them one by one. For the successfully decoded qrcode, we will append their points to the `res_points`.

So, I prefer to use the `vector`, which is more convenient to append. Could we keep it?

Reply:
> multiple QRcode in an image

So, what is represented by each vector element? 4 points corners?

---

This question is about external API representation (`points` output argument). Internal implementation can be any (we don't worry about that).

Reply:
@alalek, I suggest to merge the current PR as-is.

Reply:
@vpisarev Please make careful reviews. For example, `.fixedType()` check below doesn't make any sense. Are you going to fix that?

Reply:
@alalek  Yes, each Mat in `vector` represents 4 points corners, it can also be represented as `Mat(4, 2, CV_32FC1)`.

Reply:
Hi @alalek , the code `fixedType()` comes from https://github.com/opencv/opencv/blob/7a790d0d353bcbe48f1d62fb917c787ab4185846/modules/objdetect/src/qrcode.cpp#L49  Because I try to convert `vector<Mat>` to `OutputArrayOfArrays`. Do you have any good ideas?

Reply:
@dddzg 
Hello, are there any steps for use and deployment? I want to try to use it in Golang, but I can't run the project of wechat_qrcode. Cmake is always running unsuccessfully, and I meet all kinds of errors, which makes me very upset!

Reply:
@VladX

Reply:
Could you list the errors with more details?